### ISSUE TITLE
Allow per-hub user image customization

### ISSUE THREAD

**Issue by yuvipanda** [2i2c-org/pilot-hubs#34]
For the pilot hubs to be 'low touch', we need to not be responsible for specifying the user image. There should be a *baseline image* we maintain, with things we want to provide as 'features' to all the hubs - nbgitpuller, jupyter-tree-download, etc. We want users to be able to customize a few things on top:

1. Install new apt packages (`apt.txt`)
2. Install conda / pip packages (`environment.yml` or `requirements.txt`)
3. Install R packages (`install.R`)
4. Run arbitrary commands post installation (`postBuild`)

We *could* use [repo2docker](https://github.com/jupyterhub/repo2docker) for this, but it has a few disadvantages:

1. The images it builds are quite big
2. We can't really use newer R versions yet
3. We want to make sure that *some* packages (nbgitpuller, etc) are always installed.

The approach I prefer for this is instead what I helped do for the PANGEO images - you have a base that is maintained by us, and specific customizations via `ONBUILD`. See https://github.com/pangeo-data/pangeo-docker-images/blob/master/base-image/Dockerfile#L68 for how that would work. 

To recap, the system we want would:

1. Give us control of the base image
2. Allow hub admins to do specific additional customizations on top

We might also just allow them to use `repo2docker`, but not to begin with.

**Comment by choldgraf** [comment]
Two quick thoughts - 

1. Even if we don't use repo2docker for the builds, it would be nice if we could use the repo2docker *spec* so that the user patterns stay the same.
2. We probably want a per-hub *base image* flexibility too. Assuming the use-cases we support will diversify, there are probably going to be differences in the baseline things that are set up (e.g. a data 8 hub won't need all the stuff a pangeo hub would need, but both might be deployed at the same university)

**Comment by yuvipanda** [comment]
Totally agree on (1) - that's what the PANGEO approach is designed to do.

Agree on (2), but I'd say PANGEO styles hubs won't go out from *this* repo, but from a different hub. I think they're different enough that it'd be useful to separate them out completely.

**Comment by choldgraf** [comment]
ok that makes sense - so there is a "one configuration repo per *style* of hub" mapping, and that base configuration repo can spawn off multiple instances of that style of a hub?

**Comment by yuvipanda** [comment]
See https://github.com/2i2c-org/pilot-hubs/issues/160

**Comment by choldgraf** [comment]
Do we wanna close this and open up a new issue about a more specific enhancement we'd want to make? (e.g., some user pattern that lets them start w/ a base image and then easily build on top of it or something)

**Comment by yuvipanda** [comment]
This is done!
